packages/firestore/test/unit/specs/existence_filter_spec.test.ts

```diff
@@ -302,31 +372,124 @@ describeSpec('Existence Filters:', [], () => {
           .expectEvents(query1, { added: [docA, docB] })
           // DocB is deleted in the next sync.
           .watchFilters([query1], [docA.key], {
-            bits: { bitmap: 'AQ==', padding: 6 },
+            // Invalid 64base string in bitmap.
+            bits: { bitmap: '*#!', padding: 1 },
```

Comment:
Wow! u caught a bug. When there is an invalid character in bitmap, the decodeBase64 function will not throw error. it just skip the invalid ones.

Reply:
Manually check if the bitmap is valid base64 string, skip the validation if bitmap is UInt8Array.

Reply:
Huh you're right about decodeBase64() ignoring malformed base64:

https://github.com/firebase/firebase-js-sdk/blob/3ee35f9ae575018fa7617bd3ac69b99fbb5cc8e6/packages/firestore/src/platform/node/base64.ts#L20-L24

But it used to (here is an old commit of the same file):

https://github.com/firebase/firebase-js-sdk/blob/0b9ef3a56d8de6215ca25756f937df77959fb85a/packages/firestore/src/platform/node/base64.ts#L22-L29

Note, however, that the browser implementation of decodeBase64() calls the built-in `atob()` function which does indeed throw an exception on invalid base64:

https://github.com/firebase/firebase-js-sdk/blob/3ee35f9ae575018fa7617bd3ac69b99fbb5cc8e6/packages/firestore/src/platform/browser/base64.ts#L19-L21

https://developer.mozilla.org/en-US/docs/Web/API/atob

OK here is what I think we should do: fix the `decodeBase64()` in `src/platform/node/base64.ts` to validate, but do it in a faster way. Using a regular expression would definitely be slow. A faster way would be to verify that the length of the input string corresponds to the length of the returned byte array.

I'm thinking something like this:

```ts
class Base64DecodeError extends Error {
  readonly name = "Base64DecodeError";
}

export function decodeBase64(encoded: string): string {
  const decoded = Buffer.from(encoded, 'base64').toString('binary');
  const expectedEncodedLength = 4 * Math.ceil(decoded.length / 3);
  if (encoded.length !== expectedEncodedLength) {
    throw new Base64DecodeError(
      `invalid base64 string ` +
      `(decoded length: ${decoded.length}, ` +
      `encoded length: ${encoded.length}, ` +
      `expected encoded length: ${expectedEncodedLength})`);
  }
  return decoded
}
```

This would be orders of magnitude faster than a regular expression check and it will (I believe) catch _all_ cases of non-base64 string inputs.

See https://stackoverflow.com/questions/13378815/base64-length-calculation for details.

Reply:
I added a regular expression to check if the base64 string is valid first, then send it to decode into UInt8Array.

Reply:
Discussed offline.